pkg/storage/oss/oss.go

```diff
@@ -0,0 +1,673 @@
+		return err
+	}
+
+	d.Config = effectiveConfig.DeepCopy()
+
+	return nil
+}
+
+func fetchCredentialsIniFromSecret(secret *corev1.Secret) (auth.Credential, error) {
+	mutex.Lock()
+	defer mutex.Unlock()
+	creds, ok := secret.Data[imageRegistrySecretDataKey]
+	if !ok {
+		return nil, fmt.Errorf("failed to fetch key 'credentials' in secret data")
+	}
+
+	f, err := ioutil.TempFile("", "alibaba-creds-*")
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	defer os.Remove(f.Name())
+	defer os.Unsetenv(provider.ENVCredentialFile)
+	_, err = f.Write(creds)
+	if err != nil {
+		return nil, err
+	}
+	os.Setenv(provider.ENVCredentialFile, f.Name())
+	// use Alibaba provider initialization
+	p := provider.NewProfileProvider("default")
+	// return a valid auth credential
+	return p.Resolve()
+}
+
+func (d *driver) getCredentialsConfigData() error {
+	if d.credentials != nil {
+		return nil
+	}
+
+	// Look for a user defined secret to get the Alibaba Cloud credentials from first
+	sec, err := d.Listers.Secrets.Get(defaults.ImageRegistryPrivateConfigurationUser)
+	if err != nil && errors.IsNotFound(err) {
+		// Fall back to those provided by the credential minter if nothing is provided by the user
+		sec, err = d.Listers.Secrets.Get(defaults.CloudCredentialsName)
+		if err != nil {
+			return fmt.Errorf("unable to get cluster minted credentials %q: %v", fmt.Sprintf("%s/%s", defaults.ImageRegistryOperatorNamespace, defaults.CloudCredentialsName), err)
+		}
+	}
+
+	if sec == nil {
+		return fmt.Errorf("unable to get secret %q: %v", fmt.Sprintf("%s/%s", defaults.ImageRegistryOperatorNamespace, defaults.CloudCredentialsName), err)
+	}
+	credential, err := fetchCredentialsIniFromSecret(sec)
+	if err != nil {
+		return fmt.Errorf("failed to generate shared secrets data: %v", err)
+	}
+	// This must be an AccessKeyCrential because the oss.New requires accessKeyID and accessSecretKey
+	d.credentials = credential.(*credentials.AccessKeyCredential)
+
+	return nil
+}
+
+// getOSSRegion returns an region that allows Docker Registry to access Alibaba Cloud OSS service
+// details in https://www.alibabacloud.com/help/doc-detail/31837.htm
+func (d *driver) getOSSRegion() string {
+	region := d.Config.Region
+	return "oss-" + region
+}
+
+// getOSSEndpoint returns an endpoint that allows us to interact
+// with the Alibaba Cloud OSS service, details in https://www.alibabacloud.com/help/doc-detail/31837.htm
+func (d *driver) getOSSEndpoint() string {
+	if d.Config.EndpointAccessibility == imageregistryv1.PublicEndpoint {
+		return fmt.Sprintf("https://oss-%s.aliyuncs.com", d.Config.Region)
+	}
+	return fmt.Sprintf("https://oss-%s-internal.aliyuncs.com", d.Config.Region)
+}
+
+func (d *driver) getOSSService() (*oss.Client, error) {
+	err := d.getCredentialsConfigData()
+	if err != nil {
+		return nil, err
+	}
+
+	err = d.UpdateEffectiveConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	endpoint := d.getOSSEndpoint()
+
+	var clientOptions oss.ClientOption
```

Comment:
This is throwing a nil pointer exception.
I suggest doing something like this instead
```
	clientOptions := []oss.ClientOption{}
	if d.roundTripper != nil {
		clientOptions = append(clientOptions, oss.HTTPClient(&http.Client{Transport: d.roundTripper}))
	}
```

Reply:
got it .thank you